File: sequence-reconstruction.py
import collections

class Solution:
    def sequenceReconstruction(self, org: List[int], seqs: List[List[int]]) -> bool:
        nodes = set([j for i in seqs for j in i])
        if len(nodes) != len(org):
            return False

        neighbors = {i: [] for i in nodes}
        indegree = {i: 0 for i in nodes}
        for seq in seqs:
            for i, node in enumerate(seq[:-1]):
                indegree[seq[i+1]] += 1
                neighbors[node].append(seq[i+1])
        
        que = collections.deque([i for i in indegree if indegree[i] == 0])
        res = []
        while len(que) == 1:
            node = que.popleft()
            res.append(node)
            for neighbor in neighbors[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    que.append(neighbor)
        return res == org

File: sequence-reconstruction.java
class Solution {
    public boolean sequenceReconstruction(int[] org, List<List<Integer>> seqs) {
        Set<Integer> set = new HashSet<>();
        for (List<Integer> seq : seqs) {
            for (int i : seq) {
                set.add(i);
            }
        }
        if (org.length != set.size()) {
            return false;
        }
        Map<Integer, List<Integer>> neighbors = new HashMap<>();
        Map<Integer, Integer> indegree = new HashMap<>();
        for (int node : set) {
            neighbors.put(node, new ArrayList<Integer>());
            indegree.put(node, 0);
        }
        
        for (List<Integer> seq : seqs) {
            for (int i = 0; i < seq.size() - 1; i++) {
                neighbors.get(seq.get(i)).add(seq.get(i+1));
                indegree.put(seq.get(i+1), indegree.get(seq.get(i+1)) + 1);
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for (int node : set) {
            if (indegree.get(node) == 0) {
                queue.offer(node);
            }
        }
        
        int[] res = new int[set.size()];
        int counter = 0;
        while (queue.size() == 1) {
            int node = queue.poll();
            res[counter++] = node;
            for (int neighbor : neighbors.get(node)) {
                indegree.put(neighbor, indegree.get(neighbor) - 1);
                if (indegree.get(neighbor) == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        return Arrays.equals(res, org);
    }
}
